### PR TITLE
feat(conversation): update directive input for tool definition for model list queries

### PR DESCRIPTION
**Pull request by atierian**
## Related PR
- https://github.com/aws-amplify/amplify-api-next/pull/390

## Problem

1. Customers wanting to use generated model queries as tools for conversation routes must know the name of the generated model query in advance.

```graphql
type Todo @model @auth(rules: [{ allow: owner }]) {
  content: String
  isDone: Boolean
}

type Mutation {
  chat(
    conversationId: ID!,
    content: [ContentBlockInput],
    aiContext: AWSJSON,
    toolConfiguration: ToolConfigurationInput
  ): ConversationMessage
  @conversation(
    aiModel: "anthropic.claude-3-haiku-20240307-v1:0",
    systemPrompt: "You are a helpful chatbot. Answer questions to the best of your ability.",
    tools: [
      {
        name: "listTodos",  # <--- must match query name
        description: "lists todos",
      },
    ],
    auth: { strategy: owner, provider: userPools },
  )
}
```

2. The name of a tool must be the name of the GraphQL query. This is an arbitrary requirement we currently enforce through the directive API that isn't necessary.

## Description of changes

Updates the `ToolMap` input in the `@conversation` directive definition
```diff
  input ToolMap {
-    name: String
-    description: String
+    name: String!
+    description: String!
+    queryName: String
+    modelName: String
+    modelOperation: ConversationToolModelOperation
  }

+  enum ConversationToolModelOperation {
+    list
+  }
```

As referenced in the [inline documentation](https://github.com/aws-amplify/amplify-category-api/pull/3013/files?diff=split&w=0#diff-f57f2ca47a2e1f04b43fff35aa6b420bd0f43cf11cc3c42c392a01a2ab7c17ceR33-R38), this is essentially a fake union type.

### New DX
```graphql
# model generated list query
@conversation(
  aiModel: "anthropic.claude-3-haiku-20240307-v1:0",
  systemPrompt: "You are a helpful chatbot. Answer questions to the best of your ability.",
  tools: [
    {
      name: "list_customers",
      description: "Provides data about the customer sending a message",
      modelName: "Customer",
      modelOperation: list,
    },
  ],
  auth: { strategy: owner, provider: userPools },
)

# custom query
@conversation(
  aiModel: "anthropic.claude-3-haiku-20240307-v1:0",
  systemPrompt: "You are a helpful chatbot. Answer questions to the best of your ability.",
  tools: [
    { name: "thermometer", description: "does a thing", queryName: "getTemperature" },
    { name: "calculator", description: "does a different thing", queryName: "plus" }
  ],
  auth: { strategy: owner, provider: userPools },
)

```

### Why only support `list` queries?
- We only support queries (non-mutating) tools currently. This is an intentional constraint to prevent unexpected destructive actions through the non-deterministic nature of LLMs.
- We only support `list` queries because `get` queries require the LLM to have the primary key of the relevant model. It's possible to make this work today by including the primary key in the `aiContext` of a message, however this approach hasn't proven dependable or an acceptable DX. We're planning on exploring alternative ways to support `get` queries; until then, we're constraining model generated query tools to `list`.

### Why not just have `data-schema` generate the `queryName` rather than changing the directive API?
This was the initial route I planned, but it's the wrong one as it would force `data-schema` to take a dependency on the implementation details of the model transformer.

### What about custom named model generated list queries?
[Supported](https://github.com/aws-amplify/amplify-category-api/pull/3013/commits/0e7160dfaa95dd9169e3309533391a17350a7467) 😄 

### CDK / CloudFormation Parameters Changed
N/A

## Issue #, if available
N/A

## Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] ~E2E test run linked~
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] ~Relevant documentation is changed or added (and PR referenced)~
- [ ] ~New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies~
- [ ] ~Any CDK or CloudFormation parameter changes are called out explicitly~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by palpatim — APPROVED** [review]
Does the validation allow for mixed-mode tool maps? In other words, the map contains both a `queryName` style and a `modelOperation style?` (I can’t think of why this would be necessary, but if it’s supported, we should test it)

Approved assuming the answer is "no", or that we'll add tests for mixed-mode support later.

Also approved as API BR since this is a change to the conversation directive API.

**Comment by atierian** [comment]
> Does the validation allow for mixed-mode tool maps? In other words, the map contains both a `queryName` style and a `modelOperation style?` (I can’t think of why this would be necessary, but if it’s supported, we should test it)

A single tool definition cannot have both `queryName` (custom operation) and `modelName` / `modelOperation` (model generated query). This is validated here [[1](https://github.com/aws-amplify/amplify-category-api/pull/3013/files#diff-85826cdebb41acdefde0112b9171f40669a7de1dc7db6773054f891a2ba67966R196-R198), [2](https://github.com/aws-amplify/amplify-category-api/pull/3013/files#diff-6c75cd673bd6f6dbd63c98f54181a1800e9c2038244d5b64828a0d5e0f92ba45R25-R29)].

But a single `@conversation` can have multiple tools, which can be mixed. I thought I added a test case for that, but looking again, I only added one in [data-schema](https://github.com/aws-amplify/amplify-api-next/pull/390/files#diff-77c9378ad1789fdac5998491f0d47d14b34507ee7a84e7afa9f07fad007699c2R908-R923), which doesn't cover everything we'd get with a test here.

> Approved assuming the answer is "no", or that we'll add tests for mixed-mode support later.

If I need to make any other changes to this branch before merging, I'll include a test case for mixed tool definitions. Otherwise, I'll include it in a follow up.

File: packages/amplify-graphql-conversation-transformer/src/tools/process-tools.ts
Comment (phani-srikar) [inline packages/amplify-graphql-conversation-transformer/src/tools/process-tools.ts:270]: For constructing the list query name consistently between this and model transformer, recommend using `getFieldNameFor('list', modelName)` that is exported from the transformer core - https://github.com/aws-amplify/amplify-category-api/blob/acc8140c323988da595630ed9c047e03a825635f/packages/amplify-graphql-transformer-core/src/utils/operation-names.ts#L26

**Review by phani-srikar — CHANGES_REQUESTED** [review]

**Review by phani-srikar — APPROVED** [review]

**Comment by atierian** [comment]
The failing test is `verify_amplify_backend_compatability` because this branch isn't up to date with `main`. It's not related to these changes.

**Pull request merged**
